website/blog/2020-01-10-cobalt.md

```diff
@@ -0,0 +1,894 @@
+
+#### Restart Bloop automatically if newer version is needed
+
+Whenever Metals is started and connects to the Bloop build server it receives
+information about the version it is running. We compare that version to the one
+needed by Metals to make sure everything is properly working. If the version
+running currently is older we offer to restart it automatically and start a new
+one. Only if Bloop was installed manually it might not be work.
+
+#### Automatically reconnect to Bsp/Bloop if connection is broken
+
+There were numerous issues reported that the connection to Bps or Bloop is being
+lost, which causes Metals to behave erratically due to no compilation being run.
+To solve this, we now detect if the connection is closed and offer user the
+option of automatic reconnection. This works both when the build server was
+downed and when the connection was just interrupted. It should also solve the
+issues on MacOs with dropped connection.
+
+## Debugging
+
+It is possible now to run main methods and tests via a clickable code lens,
+however it currently only works in Visual Studio Code. We also allow setting
+breakpoints in both Java and Scala source files and to check all variables at
+each.
+
+Known breakpoints limitations we are working on:
+
+- breakpoint may get ignored when set in a type containing nested type
+  declarations preceding the breakpoint (see
+  https://github.com/scalameta/metals/issues/1195)
+- breakpoint may get wrongly resolved for an ambiguous type (e.g. a.b.Foo)
+  defined inside two modules on the project (see
+  https://github.com/scalameta/metals/issues/1196)
+
+### How to add support for debugging in my editor?
+
+You will need to handle two commands in your language client extension:
+[`metals-run-session-start`](https://github.com/scalameta/metals/blob/master/metals/src/main/scala/scala/meta/internal/metals/ClientCommands.scala#L56)
+and
+[`metals-debug-session-start`](https://github.com/scalameta/metals/blob/master/metals/src/main/scala/scala/meta/internal/metals/ClientCommands.scala#L78).
+Those commands should get executed automatically by the lsp client once the user
+activates a code lens. The difference between them is that the former ignores
+all breakpoints being set while the latter respects them. The procedure of
+starting the run/debug session is as follows:
+
+1. Request the debug adapter uri from metals server using the
+   [`debug-adapter-start`](https://github.com/scalameta/metals/blob/master/metals/src/main/scala/scala/meta/internal/metals/ServerCommands.scala#L95)
+   command.
+2. Connect the debug adapter extension specific to you editor using the
+   aforementioned uri and let it drive the run/debug session. For reference,
+   take a look at the
+   [vscode implementation](https://github.com/scalameta/metals-vscode/blob/master/src/scalaDebugger.ts)
+   and how it is
+   [wired up together](https://github.com/scalameta/metals-vscode/blob/master/src/extension.ts#L356)
```

Comment:
I think this section belong to the docs more than to the release notes.

Reply:
I agree and moved it to a separate document.